### ISSUE TITLE
Mainstreaming efforts for midas platform (N7100/N7105/I9305/I9300/I317)

### ISSUE THREAD

**Issue by tlaurion** [CustomROMs/android_local_manifests_i9300#1]
Anyone suggested the manifest to be taken by LineageOs directly so that it is automatically built, and picked up by LineageOs for microg? Can't wait to use this on a maintained OTA i9300 phone with Delta updates. Which would raise developer's attention. And keep that phone alive.

If any help is needed contact me.
Awesome work, btw.

You read that? That needs to fly and be upstreamed: https://blog.forkwhiletrue.me/posts/an-almost-fully-libre-galaxy-s3/

_Sent from my Galaxy S3 using [FastHub-Libre](https://github.com/thermatk/FastHub-Libre/)_

**Comment by ChronoMonochrome** [comment]
Hi, @tlaurion.

> Anyone suggested the manifest to be taken by LineageOs directly so that it is automatically built, and picked up by LineageOs for microg? 

What exactly do you mean? The LineageOS manifest is used in these builds being unmodified (and always a recent version). All the modifications comes to these local manifests.

>  Can't wait to use this on a maintained OTA i9300 phone with Delta updates.

IIRC this would require a server for that. Anyway if anyone have some ideas on implementing OTA updates, pull requests are always welcome.

> You read that? That needs to fly and be upstreamed: https://blog.forkwhiletrue.me/posts/an-almost-fully-libre-galaxy-s3/

Yup, before I ever started any works on S3 I began by reading these exciting articles :)

**Comment by tlaurion** [comment]
@ChronoMonochrome : i'm talking about making LineageOS build the rom themselves by asking them to pickup your manifest and changeset :)

The only reason why the exynos board phones (i9300, i9305 LTE, n7100...) were dropped from LineageOS was because [they were supported by 14.1 only](https://github.com/LineageOS/hudson/commit/b8cf20218fd7676d7904152531e0f1cacead740f) on which support was dropped following their "major version -2 support drop" which made 14.1 supported devices unsupported radically.
 
Now that you have a "working" build (frontal camera aside, on which I see that you are working on), [proposing reintegration would make sense](https://github.com/lineageos4microg/docker-lineage-cicd/issues/35#issuecomment-476692274), and will make its way into microG for LineageOS as well which create build for the same supported devices, and will also be supported by /e/.

You have complementary/oppositional information?

**Comment by ChronoMonochrome** [comment]
> @ChronoMonochrome : i'm talking about making LineageOS build the rom themselves by asking them to pickup your manifest and changeset :)

> The only reason why the exynos board phones (i9300, i9305 LTE, n7100...) were dropped from LineageOS was because they were supported by 14.1 only on which support was dropped following their "major version -2 support drop" which made 14.1 supported devices unsupported radically.

It's not that easy, I'm afraid. Device doesn't yet meet all the requirements proposed by LineageOS charter plus some hacks that are still used in the sources aren't gonna be merged easily for us. I'm not saying to stop working on exynos stuff but rather saying we are probably (likely) still not ready to claim being OFFICIAL, so need more work on it to do so.

**Comment by tlaurion** [comment]
@ChronoMonochrome Could you refer to that list?

**Comment by ChronoMonochrome** [comment]
> @ChronoMonochrome Could you refer to that list?

https://github.com/LineageOS/charter/blob/master/device-support-requirements.md

**Comment by tlaurion** [comment]
@ChronoMonochrome Don't hesitate to edit this post with references of work needing to be done after having checked everything you know meet requirements.

- [x] Audio
  - [ ]     All devices MUST support audio playback for media content.
  - [ ]     Phones MUST support in-call audio.
  - [ ]     Phones MUST support speaker audio.
  - [ ]     Tablet devices capable of in-call audio/speaker audio MUST support in-call/speaker audio.
  - [ ]     Devices SHOULD support any additional audio configuration inherent to their device (eg. echo cancellation, extra mics, etc).
  - [ ]     All devices MUST support any other audio output supported by their stock OS (eg. headphone jack, USB-C, BT).
  - [ ]     All devices with FM radio capabilities in their stock OS SHOULD support FM.

- [x] RIL
  - [ ]     All devices with RIL supported in their stock OS MUST support RIL for phone calls & data.
  - [ ]     All devices with RIL supported in their stock OS MUST support emergency calling with a SIM inserted (112/911).
  - [ ]     All devices with RIL supported in their stock OS SHOULD support emergency calling without a SIM inserted (112/911).
  - [ ]     Data-only devices (defined as devices that have a RIL but do not support telephony stack due to hardware/firmware restrictions) are EXEMPTED from phone & emergency dialing requirements.

- [ ] Encryption
  - [ ]     All devices that supported hardware-backed encryption on their stock OS MUST support hardware-backed encryption.
  - [ ]     All devices that shipped stock as forceencrypt SHOULD default to forceencrypt enabled.
  - [ ]     All devices MUST support software encryption.

- [x] Wi-Fi
  - [ ]     All devices with Wi-Fi supported in their stock OS MUST support Wi-Fi.
  - [ ]     All devices with Wi-Fi MUST report same MAC address as on stock OS.
  - [ ]     All devices with Wi-Fi hotspot capabilities MUST support Wi-Fi tethering.

- [x] USB
  - [ ]     All devices with a USB port MUST support file access via MTP.
  - [ ]     All devices with USB tethering supported on their stock OS MUST support USB tethering.
  - [ ]     All devices with a USB port & Data SHOULD support USB tethering.

- [ ] GPS
  - [ ]     All devices with GPS supported in their stock OS MUST support GPS.

- [x] Bluetooth
  - [x]     All devices with Bluetooth supported in their stock OS MUST support Bluetooth.
  - [x]     All devices with Bluetooth MUST report same MAC address as on stock OS.
  - [x]     All devices with Bluetooth SHOULD support Bluetooth tethering.
  - [ ]     All devices with support for Qualcomm® aptX™, aptX™ HD, or any future variant of aptX™, in stock (non-beta releases) OS SHOULD support those variant of aptX™.
  - [ ]     All devices without support for Qualcomm® aptX™, aptX™ HD, or any future variant of aptX™ in stock (non-beta releases) OS MUST NOT support those variants of aptX™.

- [x] Camera
  - [x]     All devices with Camera supported in their stock OS MUST support Camera, in both front facing and rear camera configurations.
  - [ ]     All devices with Dual (or more) Rear Cameras SHOULD support all rear cameras.
  - [ ]     All devices with Dual (or more) Front Facing Cameras SHOULD support all front cameras.
  - [x]     All Camera HAL versions accessible with the device's Camera HAL MUST comply with the Camera and Video Recording requirements.

- [x] Video Recording
  - [x]     All devices with Video Recording supported in their stock OS MUST support Video Recording, in both front facing and rear camera configurations.

- [x] Codecs
  - [x]     All devices with hardware encoding/decoding support in their stock OS MUST support hardware encoding/decoding for all non-proprietary codecs supported by their stock OS.

- [x] Display
  - [ ]     All devices with a built-in Display MUST support the Display at the same resolution and density as the stock OS.
  - [ ]     All devices that do not include a built-in Display MUST support Display output via the hardware’s supported outputs (eg. Android TV - HDMI).
  - [ ]     All devices that support additional non-USB display interfaces SHOULD support those display output methods.
  - [ ]     All devices that support a USB-out display in their stock OS SHOULD support this display output (eg. MHL/Miracast/OTG).
  - [ ]     All devices that support HDR10 playback in their stock OS SHOULD support HDR10 playback.

- [x] NFC
  - [ ]     All devices with NFC supported in their stock OS MUST support NFC.

- [ ] Fingerprint Sensor
  - [ ]     All devices with a Fingerprint Sensor MUST support the Fingerprint Sensor if the stock OS supports it with Marshmallow or higher Android versions.
  - [ ]     All devices with a Fingerprint Sensor SHOULD support the Fingerprint Sensor if the stock OS supports it for all other Android versions.

- [ ] IR
  - [ ]     All devices with an IR blaster SHOULD support IR blaster.

- [x] Accelerometer
  - [ ]     All devices with an accelerometer MUST support the accelerometer.

- [x] Gyroscope
  - [ ]     All devices with a gyroscope MUST support the gyroscope.

- [x] Proximity
  - [ ]     All devices with a proximity sensor MUST support the proximity sensor.

- [x] Light
  - [ ]     All devices with a light sensor MUST support the light sensor.

- [x] Other Sensors
  - [ ]     All other sensors supported by a device’s stock OS SHOULD be supported.

- [ ] Accessories
  - [ ]     All devices with proprietary accessories SHOULD support those accessories (eg. O-Click, Essential 360 Camera).

- [ ] Hardware Deviations 
  - [ ] Hardware deviations are defined as exemptions granted for hardware requirements above that worked in stock, but do not work in LineageOS.
  - [ ]     All hardware deviations from stock MUST be reported on the Wiki page for the device, with a user understandable justification.

- [ ] Software support
  - [x] Device tree structure
    - [ ]     Device trees MUST contain a Lineage-specific makefile with device declaration of lineage_[devicename].
    - [ ]     Device trees MUST support a lineage.dependencies file for breakfast command & roomservice to be functional.
    - [ ]         This file MUST NOT include any dependencies outside of the "LineageOS" organization.

- [x] Build type
  - [ ]     All devices MUST be configured as userdebug releases.

- [ ] Kernel
  - [ ]     All devices MUST NOT ship a prebuilt kernel.
  - [ ]     All devices MUST NOT implement software based touchscreen wake features such as double tap to wake, swipe to wake or gestures if there is no hardware-backed support for them in the touchscreen firmware.
  - [ ]     All devices MUST NOT implement forced fast charge over USB methods that violate the USB specifications.
  - [ ]     All devices MUST NOT implement any form of clock manipulation (underclocking, overclocking, etc.) for any processor (CPU, GPU).
  - [ ]     All devices MUST NOT implement any form of hardware voltage manipulation (undervolting, custom voltage tables, etc.).
  - [ ]     All devices MUST NOT implement any form of hardware register manipulation (sound control, etc.).
  - [ ]     All devices MUST NOT implement any form of custom KSM driver (UKSM, etc.).
  - [ ]     All devices MUST NOT ship governors other than the ones specified in the following list:
    - [ ]         conservative
    - [ ]         interactive
    - [ ]         ondemand
    - [ ]         performance
    - [ ]         powersave
    - [ ]         sched
    - [ ]         schedutil
    - [ ]         userspace
  - [ ]     All devices MUST NOT ship I/O schedulers other than the ones specified in the following list:
    - [ ]         bfq
    - [ ]         cfq
    - [ ]         deadline
    - [ ]         noop
    - [ ]         row
  - [ ]     All devices MUST only ship hotplugging drivers provided by the OEM or SoC vendor.

- [ ] SELinux status
- [ ]     All devices MUST be configured for SELinux Enforcing.

- [x] Verity
  - [ ]     All devices MUST disable verity on the system image for userdebug builds.
  - [ ]     All devices SHOULD support verity on the vendor image.

- [ ] Updater
  - [ ]     All devices with a shipping build of LineageOS MUST support upgrades via the native LineageOS Updater application & the recovery documented on the Wiki for that device.

- [ ] FRP
  - [ ]     All devices with stock support of Factory Reset Protection (FRP) SHOULD support FRP when Google Applications are installed by the user.

- [x] SafetyNet
  - [ ]     All devices MUST NOT alter SafetyNet validation responses.

- [x] Binder
  - [ ]     All devices MUST use the 64-bit Binder API.

- [x] Root (su) 
  - [ ]     All devices MUST NOT ship with su included.
  - [ ]     All devices MUST support su installation via LineageOS provided ‘Extras’ download.

- [x] Non-PIE Blobs
  - [ ]     Devices MUST NOT use non-PIE binaries.

- [ ] Proprietary files extraction
  - [ ]     Devices MUST have a working proprietary files extraction script in their device tree (or device tree dependencies) that reproduces an exact copy of the binaries required to build LineageOS from an existing LineageOS installation.
  - [ ]     Devices SHOULD use the global extraction script (located in vendor/lineage).
  - [ ]     If a device maintainer elects to not use the common extraction script, the maintainer MUST ensure that the Wiki page for their device has valid instructions for operating the custom extraction script.

- [x] CVE
  - [ ]     Devices MUST support CVE patches for “high profile” exploits and vulnerabilities (if the media is reporting on it, then we must have it patched).
  - [ ]     [NOTE: This will become a MUST once CVE autopatcher is live & automated]
  - [ ]     Devices SHOULD receive regular CVE patches to the device kernel and dependencies.
  - [ ]     [To be in effect once CVE autopatcher is live & automated]
  - [ ]     Device maintainers MUST review and/or accept patches provided by the CVE autopatcher tool.

- [x] Firmware Assert
  - [ ]     All devices MUST assert on known to be working firmware versions if some firmware versions are known to be non-working.

- [ ] exFAT Support
  - [ ] LineageOS operates under the assumption that OEM device licensing for exFAT is attached to the device, not software. LineageOS will comply with all requests for removal of exFAT support from OEMs, Microsoft or their representatives upon contact to legal@lineageos.org.

    - [ ]     All devices with exFAT support on stock MAY support exFAT with (and only with) a kernel based implementation.
    - [ ]     All devices without exFAT support on stock MUST NOT support exFAT.

- [x] Additional Features 
  - [ ]     All devices SHOULD support in-kernel (MDSS, MDNIE or similar) LiveDisplay colour adjustment.

- [ ] Software Deviations
  - [ ] Software deviations are defined as exemptions granted for software requirements above that worked in stock, but do not work in LineageOS.
    - [ ]     All software deviations from other LineageOS devices of the same type MUST be approved by Directors (eg. if one wants to remove Music app, get approval).
    - [ ]     All software deviations from other LineageOS devices of the same type MUST be reported on the Wiki page for the device, with a user understandable justification.
    - [ ]     Device maintainers MUST ship Jelly or another LineageOS sourced web browser.

- [x] Quality of life
  - [ ] Commit Authorship
    - [ ]     All non-original commits MUST have proper authorship attribution from the source it was taken from or adapted from.

- [x] Copyrights
  - [ ]     All original contributions MUST be copyrighted as “(C) [YEAR] The LineageOS Project”.
  - [ ]     All LineageOS copyrights MUST only be additive to the copyright header.
  - [ ]     Do not remove copyrights from CyanogenMod, Cyanogen Inc or any other upstream.

- [x] Workflow
  - [ ]     Force pushing branches SHOULD be avoided.
  - [ ]     In the event of a force pushed branch, backup branches of the pre-forced HEAD MUST be made.

- [ ] JIRA
  - [ ]     Device maintainer(s) MUST have a JIRA account for bug tracking and cross-team collaboration.
  - [ ]     Device maintainer(s) MUST routinely triage, answer and close JIRA reports.
  - [ ]     Device maintainer(s) SHOULD make their JIRA name match their maintainer name as displayed on the Wiki.

- [x] Licensing
  - [ ]     All Kernel contributions MUST be GPLv2.
  - [ ]     All Android contributions SHOULD be Apache 2.0 licensed.
  - [ ]     Any contribution to an existing Apache 2.0 project MUST fall under Apache Compliance Category A.
  - [ ]     Any contribution to an existing Apache 2.0 project MUST NOT be in Apache Compliance Category X.

- [x] Wiki
  - [ ]     All devices with a shipping build of LineageOS MUST have a Wiki page with valid installation instructions.
  - [ ]     All devices with a shipping build of LineageOS MUST document Hardware Deviations from stock capabilities.
  - [ ]     All devices with a shipping build of LineageOS MUST document Software Deviations from other LineageOS releases of the same device type.

- [ ] Stability
  - [x]     Issues like the "screen of death" MUST NOT affect the device.
  - [ ]     The device MUST NOT have abnormal battery drain.
 
- [x] Recovery
  - [ ]     Maintainers MUST document for users on the Wiki a valid Recovery image by which to install LineageOS zip files.
  - [ ]     Devices that do not have traditional Recovery images MUST support & document another means of installation for LineageOS zip files.
  - [ ]     Maintainers SHOULD verify that Teamwin Recovery Project (TWRP) official distributions work for LineageOS installation.
  - [ ]     Failures in official TWRP recoveries should be raised with the TWRP team or remedied by the maintainer.
  - [ ]     Maintainers SHOULD provide a custom recovery link in Wiki documentation if TWRP does not officially support their device.

**Comment by ChronoMonochrome** [comment]
@tlaurion, nice idea! I thought of creating a fork of the charter repo to indicate the progress being done, though.

**Comment by tlaurion** [comment]
@ChronoMonochrome : Pinpoint where it's at! Interested in making this go forward. 

Could do a [CI](https://github.com/lineageos4microg/docker-lineage-cicd/issues/35#issuecomment-476692274) and modify [LineageOS Updater](https://github.com/LineageOS/android_packages_apps_Updater) url to point to the [updater backend](https://gitlab.com/LineageOS/infra/updater)  on AWS (or whatever free server recommended) for the pre-official community to test this properly?

Never did that before but it seems to be the way to go! What do you think?

**Comment by ChronoMonochrome** [comment]
> Could do a CI and modify LineageOS Updater url to point to the updater backend on AWS (or whatever free server recommended) for the pre-official community to test this properly?

> Never did that before but it seems to be the way to go! What do you think?

Yes, that should do. Though I have no clues about which free servers can be suitable for LineageOS updater.

**Comment by pgarrity18** [comment]
Apologize if this is a dumb question...

> And there we have it: we've successfully removed almost all blobs from the Exynos 4412 found in the i9300. The one blob remaining is probably not replaceable: it's both encrypted and signed by Samsung. However, since we have access to where the code is stored, it would be trivial to wipe every trace of it from RAM once U-Boot starts. (And similarly, it's easy to dump the code that's been loaded into RAM - it is not even remotely interesting...)

forkbomb calls this approach "almost fully libre". How does this differ from the approach used to [deblob and remove the Intel ME on GM45 chipsets](https://libreboot.org/docs/hardware/gm45_remove_me.html)? Is this something that could ultimately get fsf RYF certification? Or is it just really close? 

Just trying to understand how significant this is. Should I order my Note 2 before they all get scooped up? Thanks guys!

**Comment by tlaurion** [comment]
@pgarrity18:

TLDR: i9300 won't get RYF certification, for the same reason the Lenovo x220 didn't under minifree attempt, and why Librem laptops won't ever receive it: the second CPU still runs binary blobs.



From my understanding, that would not get RYF certification, just like Ivy bridge and Sandy bridge based platforms can't, since Intel ME is still present, boots, and pass control to main CPU, but without having any interface to interact with it from a booted system. 


RYF certification requires "complete removal" aka disablement. GM45 permitted that. 


For Ivy and Sandy bridges, we talk about neutering, since only ROMP and BUP modules are required to pass control to the main CPU.


For later bridge implementation in platforms, we talk about disablement, since ME is "asked" to disable itself from discovered hidden instructions from positive research, while still requiring a bunch of other modules to be present in ROM to pass control to main CPU. They cannot be removed, me_cleaner can't trim the space. They are there.

In current case, what @fourkbomb means there is that a complete bypass is possible, since there is a kexec happening, from which all memory mappings can be wiped.


There is a possibility to remove any trace in memory of that blob, and that the other CPU having passed control doesn't know that its access had been revoqued. But that needs to be tested thoroughly. 


Note the absence of further posts on that blog. Would be nice if @fourkbomb replied to contact attempts. But didn't.

_Sent from my Galaxy S3 using [FastHub-Libre](https://github.com/thermatk/FastHub-Libre/)_

**Comment by pgarrity18** [comment]
Thank you for the response. It is interesting indeed and I too wish that forkbomb was more engaged with the community. I am trying to wrap my brain around what he did, what these guys did ..

https://blog.fossencdi.org/u-boot-galaxys3.html, 

and what other possibilities might still exist. For example, could you repackage the S3 after shorting the bridge and forcing a boot from the SD card via u-boot? Even without RYF certification, the work you guys are doing is exceptional!

**Comment by tlaurion** [comment]
@pgarrity18 
> https://blog.fossencdi.org/u-boot-galaxys3.html,

That is interesting. Will read thoroughly. I do not take any credit on this. But making QubesOS preinstalled on Heads booted open source and QubesOS certified hardware, I do :) Stay tuned!

See state of [ME on ivy bridge](https://github.com/osresearch/heads-wiki/blob/master/Clean-the-ME-firmware.md)
See required [coreboot blobs](https://github.com/osresearch/heads/tree/master/blobs) for X220, Librems

**Comment by tlaurion** [comment]
@fourkbomb: can you help in accomplishing what is missing to mainstream this?

**Comment by tlaurion** [comment]
Attempted @fourkbomb contact by email. 

_Sent from my Galaxy S3 using [FastHub-Libre](https://github.com/thermatk/FastHub-Libre/)_

**Comment by herbsmn** [comment]
The Replicant project is working to port the i9305/i9300 to LineageOS 16.1 using the mainline Linux kernel. Help is wanted. https://redmine.replicant.us/projects/replicant/wiki/Porting_Replicant_to_Android_9/#LineageOS-16

Hit us up on IRC or our email list if you want to join in the fun: https://redmine.replicant.us/projects/replicant/wiki#Contact

**Comment by herbsmn** [comment]
Commits are happening here: https://git.replicant.us/contrib/GNUtoo/device_i9305.git/log/?h=GNUtoo/lineage-16.0-base

**Comment by ChronoMonochrome** [comment]
@herbsmn, it's nice to see other people also working on Midas! I've attempted to build LineageOS 15.1 with Lima userspace libs, but shortly gave up.
Will try out a Replicant OS build to see if I can get it work on I9300.

**Comment by herbsmn** [comment]
@ChronoMonochrome are you able to jump in #Replicant on Freenode's IRC to chat with us as you go? ping GNUtoo, Putti, or sensiblemn if you do.

**Comment by ChronoMonochrome** [comment]
@herbsmn, I've successfully built a i9300 ROM and got to the bootanimation (as article states, it's very slow). Is there any interest to work on it until graphics issues are resolved? I gave another try to Lima driver, but it seems like Lima userspace lib doesn't implement DRI API methods thus it fails to load.

**Comment by herbsmn** [comment]
The Replicant project has a large interest on continuing to work on mainline LineageOS on midas for a long time. We also plan on replacing the proprietary s-boot with mainline u-boot. Anytime you want to jump in and help out, we would be glad to collaborate with you.

**Comment by ChronoMonochrome** [comment]
Okay, if I'll get anything useful, I'll contact. Though I'm not sure if I'm able to fix graphics issue, which seems to be the most crucial (unless folks supporting Lima can fix it), but I'll give a try.

**Comment by herbsmn** [comment]
Sounds great! Here's a bit of an outline on our plan of attack. https://redmine.replicant.us/projects/replicant/wiki/TasksFunding#Graphics-acceleration Any questions or comments you might have would be very welcome.

**Comment by herbsmn** [comment]
Graphics are now working, but it is very very slow. We were able to get to the homescreen though and we confirmed that the touchscreen is working. We now will be working on graphics acceleration, but perhaps other work can be done now too even though the graphics are super slow.

**Comment by herbsmn** [comment]
Replicant just had a bit of a breakthrough today on the graphics using a somewhat big hack that needs to be fleshed out. You can see the speed of our new build with the hack on the devices on the left of the videos here, with the devices on the right being what we were previously experiencing: https://grimkriegor.zalkeen.net/public/alaunch.mp4 https://grimkriegor.zalkeen.net/public/slides.mp4 We hope to be pushing code that impliments this new user experience soon, so that anyone can build it.

If anyone wants to help us run LineageOS 16 with a mainline kernel and a mainline bootloader (with proprietary s-boot and proprietary TrustZone removed), with nearly zero proprietary blobs you can contact us here: https://redmine.replicant.us/projects/replicant/wiki#Contact

Documentation on the porting work to Android 9 for the i9300 (and other midas devices) can be found here: https://redmine.replicant.us/projects/replicant/wiki/Porting_Replicant_to_Android_9

**Comment by ChronoMonochrome** [comment]
@herbsmn, 

wow, such a nice progress. I was trying to bring up either Lima or Mali drivers to the mainline kernel, but didn't got any progress so far.
I'm still quite busy with other projects, but I'll rebuild it as soon as I have enough time.

**Comment by tlaurion** [comment]
Awesome guys. Have a n7100, no i9305 yet. Will contact once I equipped myself. Thanks for your awesome work guys!

**Comment by tlaurion** [comment]
Hmmm. Is it a n7100 I see in the video? :grinning:

**Comment by dllud** [comment]
@tlaurion it is two i9305 in the video.

**Comment by herbsmn** [comment]
@ChronoMonochrome @tlaurion glad to hear that you two are interested in Replicant's progress and want to join in the fun! here's our blog post about the awesome work that @dllud and others are doing on the graphics side of things: https://blog.replicant.us/2019/07/graphics-support-for-replicant-9/

**Comment by herbsmn** [comment]
@ChronoMonochrome I noticed someone over at the ODROID forums tested the lima driver on an Exynos4412 dev board with kernel 5.3-rc3 and it didn't result in a fatal error: https://forum.odroid.com/viewtopic.php?f=55&t=3691&sid=f73a6ad3052ead98bb90fa7854d061db&start=400#p264856

**Comment by ChronoMonochrome** [comment]
@herbsmn, 
AndroX (Note 2 developer) claimed that he got Lima drivers running on 4.9 (with a huge support from Simon and other people from Code Aurora Forum). There are several screenshots posted in Note2 Telegram group:
<details>
 <summary>Screenshots</summary>

![Screenshot from 2019-08-22 20-39-00.png](https://i.loli.net/2019/08/23/rxM4fpeoCXbDjTU.png)
![photo6249219910542666525.jpg](https://i.loli.net/2019/08/23/FYhuRn5KfqgVAe6.jpg)
![Screenshot from 2019-08-22 20-37-39.png](https://i.loli.net/2019/08/23/uilEKgXMn5dB3cV.png)

</details>

Unfortunately, for now no sources are available publicly, and even not clear whether they are going to be released (at least not earlier than first test builds are released).
Maybe someone could ping those people from Code Aurora Forum or Simon for getting some help on this.

About Lima driver:
I don't really know what is so special about LK 4.9. I've tried myself running 4.9, it seems that it boots but without any graphics. The oldest mainline-ish kernel that still has a working DRM subsystem for me was 4.11 (with a several backports from 4.12). I don't know if that makes sense to go deeper and trying to lower the kernel version to match exactly 4.9 (it will still have to rely on 4.12 backports in order to get DRM working).

By the way, some months ago I found that Samsung has a kernel source code for their Tizen OS (which is almost mainline, the newest version I could find was 4.1): 
https://git.tizen.org/cgit/platform/kernel/linux-exynos/log/?h=tizen_5.0

It has Mali graphics support integrated, I have some suspects that the platform specific changes may be useful for Lima as well. By itself, LK 4.1 doesn't boot on ReplicantOS (it seems to be the same issues with DRM subsystem), so rebasing changes on top of at least 4.4 or 4.9 is needed (I will try that if I don't succeed with Lima bringup on LK 5.2 / 5.3).

EDIT:
Gave it a more attempts to boot up with a different kernel version.
On 5.3-rc5 I have:
```
[    0.872750] exynos-dsi 11c80000.dsi: failed to get regulators: -517
[    0.875231] lima 13000000.gpu: bus rate = 24000000
[    0.875263] lima 13000000.gpu: mod rate = 50000000
[    0.875313] lima 13000000.gpu: failed to get regulator: -517
[    0.875341] lima 13000000.gpu: regulator init fail -517
[    0.875371] lima 13000000.gpu: Fatal error during GPU init
```

4.17 is a bit more informative:
```
[    3.123287] exynos-dsi 11c80000.dsi: Failed to get supply 'vddcore': -517
[    3.130099] exynos-dsi 11c80000.dsi: failed to get regulators: -517
[    3.138778] lima 13000000.gpu: bus rate = 160000000
[    3.142192] lima 13000000.gpu: mod rate = 24000000
[    3.146982] lima 13000000.gpu: failed to get regulator: 0
[    3.152374] lima 13000000.gpu: regulator init fail -517
[    3.157556] lima 13000000.gpu: Fatal error during GPU init
```

4.13 and 4.11 just hangs so I couldn't get any logs here.

For some reason the voltage regulator is not initialized.
I'm quite new to this stuff, yet googling for a similar issues didn't give any results so far.
Maybe checking how the stock kernel initializes this regulator can give some hints why the mainline kernel fails to do so.

**Comment by ChronoMonochrome** [comment]
Lima kernel driver is now loading with hacks:
https://github.com/ReplicantOS-midas/android_kernel_samsung_smdk4412/commits/master

```
08-23 05:08:05.858     0     0 I exynos-dsi 11c80000.dsi: failed to get regulators: -517
08-23 05:08:05.860     0     0 I lima 13000000.gpu: bus rate = 24000000
08-23 05:08:05.860     0     0 I lima 13000000.gpu: mod rate = 50000000
08-23 05:08:05.860     0     0 E lima 13000000.gpu: failed to get regulator: -517
08-23 05:08:05.861     0     0 I lima 13000000.gpu: gp - mali400 version major 1 minor 1
08-23 05:08:05.861     0     0 I lima 13000000.gpu: pp0 - mali400 version major 1 minor 1
08-23 05:08:05.861     0     0 I lima 13000000.gpu: pp1 - mali400 version major 1 minor 1
08-23 05:08:05.861     0     0 I lima 13000000.gpu: pp2 - mali400 version major 1 minor 1
08-23 05:08:05.861     0     0 I lima 13000000.gpu: pp3 - mali400 version major 1 minor 1
08-23 05:08:05.861     0     0 I lima 13000000.gpu: l2 cache 128K, 4-way, 64byte cache line, 64bit external bus
08-23 05:08:05.862     0     0 I         : [drm] Initialized lima 1.0.0 20190217 for 13000000.gpu on minor 0
...
08-23 05:08:05.973     0     0 E regulator_load_fn: init
08-23 05:08:05.973     0     0 E regulator_load_fn: enabled DSI regulator, ret=0
...
```
It seems to be ok, according to kernel log, this voltage regulator can be loaded but needs a time to initialize (and it's initialized at least without any error codes).

But now there is another problem, it doesn't look like drm hwcomposer is compatible with Lima:
```
08-23 05:08:14.964   331   331 E hwc-drm-device: Failed to set universal plane cap -1
08-23 05:08:14.964   331   331 E hwc-resource-manager: Failed to initialize any displays
08-23 05:08:14.964   331   331 E hwc-drm-two: Can't initialize the resource manager -22
08-23 05:08:14.964   331   331 E hwc-drm-two: Failed to initialize DrmHwcTwo err=6
08-23 05:08:14.964   331   331 E ComposerHal: failed to open hwcomposer device: Invalid argument
08-23 05:08:14.965   331   331 E android.hardware.graphics.composer@2.1-service: Could not get passthrough implementation for android.hardware.graphics.composer@2.1::IComposer/default.
```

In the case if anyone interesting, here is the full boot log:
https://gist.github.com/ChronoMonochrome/3ea0ab3bfc6677f412810e100ac9a6da

**Comment by herbsmn** [comment]
Very exciting! Thanks for sharing!

I'm curious, are you using freedesktop's version of drm hwcomposer? https://gitlab.freedesktop.org/drm-hwcomposer/drm-hwcomposer

**Comment by dllud** [comment]
Great news @ChronoMonochrome !

It is normal for drm_hwcomposer to not work with Lima. AFAIK Lima should only expose a render node (the Mali GPU is not connected to a display). You will have to use drm_hwcomposer with the master node from drm/exynos. drm/exynos and drm/lima can then share buffers through PRIME.

**Comment by ChronoMonochrome** [comment]
@herbsmn,
Yes, I'm using the same drm hwcomposer as posted in the ReplicantOS manifest (so it's freedesktop version).
I had to update mesa3d repo and rebase a few patches on top of it (building with original repo was giving errors with Lima enabled).

@dllud, 
oh, seems to be complicated. I have left DRM_EXYNOS enabled in kernel, but I guess, some extra changes are still needed in drm_hwcomposer?

**Comment by dllud** [comment]
It should just be a matter of pointing drm_hwcomposer to the correct dri node on your system (probably `/dev/dri/card0`). You can take a look at our [system.prop](https://git.replicant.us/contrib/replicant-9/device_samsung_i9305/tree/system.prop#n21).

I haven't tried the buffer sharing part yet, so I am helpless at that. Before diving into Lima we will first try to have drm_hwcomposer and gbm_gralloc working properly with DRM Magic (more info is available at the Replicant wiki: [Graphics on Replicant 9](https://redmine.replicant.us/projects/replicant/wiki/GraphicsReplicant9)).

PS. If you feel that sharing ideas live will help you, please drop by at Replicant's IRC channel on Freenode (#replicant).

**Comment by herbsmn** [comment]
Here's some pictures of a person successfully running Android on MALI-400 using lima with a Orange Pi Plus 2E: https://gitlab.freedesktop.org/lima/mesa/issues/120

**Comment by ChronoMonochrome** [comment]
> Here's some pictures of a person successfully running Android on MALI-400 using lima with a Orange Pi Plus 2E: https://gitlab.freedesktop.org/lima/mesa/issues/120

These are exciting news to hear!
Finally I was myself able to get Lima running on S3 too (thanks to changes proposed in OrangePi Plus 2E repos), have a look at https://github.com/ReplicantOS-midas (had to update mesa3d, device tree and kernel).
It's running a somewhat better than with a software rendering (still it could be probably much faster, because for now Mali seems to be running at 50 MHz).
Having the same issues described in the link above (graphic artifacts and freezes), gonna try a fix proposed.

**Comment by herbsmn** [comment]
This is all very exciting indeed! Thanks for the news! Looking forward to hearing if the proposed fix for the artifacts and freezes worked for you.

**Comment by ChronoMonochrome** [comment]
@herbsmn, 

applying the suggested fix indeed fixed some textures being not rendered (however, some UI elements are still missing or rendered not properly).
Freezes seems to have a different nature. It's quite difficult for me to get any logs at this point, because when it freezes, the ADB connection drops as well, and after a hard reboot it's impossible to get logs from a previous boot.
So far I was able to get just one log (https://gist.github.com/ChronoMonochrome/5e72ed87061f46539863c1fd572ebd63) which seems to be related to some sort of memory allocations issues.
I've applied a patch that hopefully should resolve memory allocation issues - https://github.com/ReplicantOS-midas/android_kernel_samsung_smdk4412/commit/0a1164f2400d1f8c3f3e84aa18b28e4452deeeac (I can't recall right now where exactly to find an original patch, but this idea of hacking the MAX_ORDER was present in some of Samsung kernel sources).

Still having a frequent freezes though (and sometimes they occur even when issuing a dmesg or logcat in ADB - I've also took USB configs from Orange PI Plus 2E device tree, but have no clues if it's anyhow a cause of freezes).

**Comment by fourkbomb** [comment]
As far as I know, lima should be able to use non-contiguous memory on exynos. Make sure you have Exynos IOMMU enabled in the kernel? You might also have to tweak the way gralloc allocates memory to get it working.

**Comment by ChronoMonochrome** [comment]
@fourkbomb, thanks for participation in a discussion!

> Make sure you have Exynos IOMMU enabled in the kernel? 

Yeah, I think it should be already enabled. 
I might've wrongly assumed that memory allocation issue I faced was the only problem. There probably should be something else causing UI freezes. Anyway I'll check how frequently allocation errors appears in log (without MAX _ORDER hack).

> You might also have to tweak the way gralloc allocates memory to get it working.

Umm, honestly, I don't know what exactly to look for. Anyway I'll try to log how much memory it needs to allocate at once to see if we indeed exceeding the limit set by MAX_ORDER.

**Comment by herbsmn** [comment]
Here's a video of lima running on Replicant 9: http://www.belg.in/replicant_9.webm

**Comment by javifo** [comment]
Wow! I'm impressed with the progress!

I would like to build it and test it. I have to organize me to get a couple hours/week free to fix things. I'm interested mainly in gralloc/hwcomposer.

**Comment by javifo** [comment]
Lima driver do allow non-cma memory

https://gitlab.freedesktop.org/lima/mesa/issues/29

**Comment by herbsmn** [comment]
Hi @ChronoMonochrome. I just wanted to check in to see if you've been working on this at all.

**Comment by ChronoMonochrome** [comment]
Hi, @herbsmn

Last time I attempted using LOS16.0 on a mainline (it was about month ago), I can say it's already much better than it was before (I got rid of kernel crash / freeze issue, I think it was totally unrelated to Lima).
There's also some progress implementing audio support (again, only thanks to @fourkbomb work).
Managed to get the speaker to work (but that's all, neither headphones nor microphones do work). 
I switched for now to the non-mainline kernel because of issues I can't solve (Lima still has crashes in userspace + audio-related issues). I'm planning to check soon if the headset can be fixed by using a proper audio policy config.
I'll leave the link to the audio patches in the case anyone can find it useful: https://github.com/ReplicantOS-midas/android_kernel_samsung_smdk4412/commits/lk-5.3

**Comment by herbsmn** [comment]
Hi @ChronoMonochrome,

Replicant would love to work more collaboratively on this project with you, if you'd be interested. Please don't hesitate to pop into our IRC or email list from time to time to provide us with details of small victories or frustrating problems. Otherwise, at least two of us are following this github thread if you prefer to post updates here. Thanks so much for continuing this important endevour! Cheers!

**Comment by ChronoMonochrome** [comment]
@herbsmn  

I was planning to post on IRC that aforementioned kernel freezes are resolved now
(but since it appeared they aren't related to the Lima at all, but rather caused by my mods to the device tree, I thought that doesn't even worth mentioning).
I don't know if I will return to the mainline kernel any soon, but if I'll find anything that worth any interest (either it's a win or a fail), I'll post it on IRC.

**Comment by Viciouss** [comment]
Hey there, I'm very happy I stumbled upon this thread, a lot of useful links to read through.

I'm quite new to the kernel world and Android (like 6 months) and still learning new things every day but I try to work my way through all of this so that I can run a mainline kernel on my galaxy note 10.1, which is very similar to the i9300. I'm able to boot into buildroot with a 5.4 kernel, display, touch, wifi and some other stuff already working. I can run kmscube but I think it's not using the lima driver but software rendering as it's really slow. I'm also setting up an Android build (10 right now) in parallel with drm hwcomposer but I'm stuck at selinux context initialization in the init program so no positive boot yet.

I wanted to thank you all for your great work on the exynos4 as I found a lot of resources all over the place which helped me to grow my knowledge bit by bit and especially @fourkbomb because your blog motivated me to look beyond my own nose and do something different from my 9-to-5 java employment in the first place.

**Comment by herbsmn** [comment]
This is very exciting news, @Viciouss! We at Replicant are extremely interested in hearing more details about this. Any chance you'd be willing to join our email list or IRC channel and chat with us a bit more about it, or alternatively post a link to a blog post you've written about your journey? Posting more details here is quite alright as well, since a number of us are following this thread. Congrats on these hacking victories and thanks for sharing!

**Comment by herbsmn** [comment]
If anyone is reading this and wants to help @Viciouss hack on the Note 10.1 (2012 edition), here's a list of the device names for the Exynos4412 Note 10.1 that you would need: https://redmine.replicant.us/projects/replicant/wiki/Exynos4412Devices#Galaxy-Note-101-2012-Edition

**Comment by herbsmn** [comment]
There's some recent work being done to mainline the ODROID U3, which has the Exynos4412 SoC, here: https://github.com/hexdump0815/linux-mainline-and-mali-on-odroid-u3/releases

**Comment by herbsmn** [comment]
Hey @hexdump0815, we would love to coordinate closer with you to mainline Exynos4412 devices. Please feel free to post about your progress in this thread.

**Comment by hexdump0815** [comment]
hello,

thanks for letting me know about the project - i think i simply do not have the time to get deeper into this project as well - feel free to take any info you can find here https://github.com/hexdump0815/linux-mainline-and-mali-on-odroid-u3 if there is anything which might help you.

good luck for your project and best wishes - hexdump

**Comment by ChronoMonochrome** [comment]
Hey @herbsmn

Any progress here with the mainline on Exynos 4?
Just recently tried to build Replicant 9 with the mainline kernel once again on i9300.
Still struggling trying to make audio work with headphones.
It's working somewhat better with the devfreq enabled (had to remove some OPP steps though).
There is another issue, for some reason the deep sleep is not working.
It would be almost usable already if those issues could be resolved.

https://github.com/ReplicantOS-midas/android_kernel_samsung_smdk4412/commits/lk-5.3

Some people are trying to bring mainline on i9100 as well:

https://github.com/Sekilsgs2/i9100_kernel_mainline_port/issues/1

Reportedly, USB OTG is working here as well (only on LK 4.20 though), which seems to be broken on i9300.

**Comment by dllud** [comment]
Hey @ChronoMonochrome , thanks for keeping up the effort.

> It's working somewhat better with the devfreq enabled

Ohh, [we had to disable devfreq](https://git.replicant.us/contrib/replicant-9/kernel_replicant_linux/commit/?id=37f3a370a08df148dc1f26f1ac1cf52cc7bb75cd) due to [glitches on graphics display when using hardware planes with software rendering](https://ftp.osuosl.org/pub/replicant/conferences/replicant-contributors-meeting-july-2019-france/graphics-acceleration-on-replicant.pdf). You get no such issue with Lima?

We will try it with [your patch to remove the OPP steps](https://github.com/ReplicantOS-midas/android_kernel_samsung_smdk4412/commit/ce08237e39e2d74c1a25b3eadde4ac0f1043f270).

Anyhow, do you have any idea on why the headphones behave better with devfreq enabled?

**Comment by ChronoMonochrome** [comment]
@dllud

> Ohh, we had to disable devfreq due to glitches on graphics display when using hardware planes with software rendering. You get no such issue with Lima?

I guess I saw somewhere in the mailining lists an attempt to remove lower freq OPP steps (but it was declined, probably there are some devices that can use those low freq OPP steps without issues). 
I'm not 100% sure that removing OPP steps is still necesary, probably I want to re-test without that patch as well. But there are no issues with this patch currently.


UPD.
>  glitches on graphics display when using hardware planes with software rendering

 I didn't test the software rendering, but currently with Lima there were no issues (and hardware planes are enabled).

> Anyhow, do you have any idea on why the headphones behave better with devfreq enabled?

Sorry for misunderstanding, headphones doesn't currently work, it's not related to the Busfreq (I should have mentioned that), it just seems that the CPU temperature was lower with the devfreq enabled.

**Comment by herbsmn** [comment]
Hi @ChronoMonochrome,

Replicant has been spending a fair bit of time working on getting a new 6.0 release out the door, so I don't know if there is much to report on Replicant 9. I think a number of us will be transitioning back to Replicant 9 next week. Thanks for checking in though and letting us know about the i9100 work that is being done!

Does anything on this page help at all with the headphone issue you are experiencing? https://wiki.postmarketos.org/wiki/Audio

**Comment by ChronoMonochrome** [comment]
@herbsmn 

Thanks for the link, have already checked that one.
I think the problem lies somewhere in kernel (DTS?), as it seems like the sound support for wm1811 is not fully implemented. Below is the dmesg log from LK 5.3:

```
[    1.725996] midas-audio sound: GPIO lookup for consumer fm-sel
[    1.726026] midas-audio sound: using device tree for GPIO lookup
[    1.726120] midas-audio sound: GPIO lookup for consumer lineout-sel
[    1.726143] midas-audio sound: using device tree for GPIO lookup
[    1.726241] midas-audio sound: using lookup tables for GPIO lookup
[    1.726264] midas-audio sound: No GPIO consumer lineout-sel found
[    1.726434] midas-audio sound: ASoC: CODEC DAI wm8994-aif1 not registered
[    1.726463] midas-audio sound: Failed to register card: -517
[    1.985821] midas-audio sound: ASoC: CODEC DAI wm8994-aif1 not registered
[    1.985878] midas-audio sound: Failed to register card: -517
[    2.245972] midas-audio sound: ASoC: CODEC DAI wm8994-aif1 not registered
[    2.246019] midas-audio sound: Failed to register card: -517
[    2.828739] midas-audio sound: wm8994-aif1 <-> samsung-i2s mapping ok
[    2.873809] midas-audio sound: wm8994-aif2 <-> samsung-i2s mapping ok
[    2.894806] midas-audio sound: wm8994-aif3 <-> samsung-i2s mapping ok
[    2.926638] midas-audio sound: Connecting non-supply widget to supply widget is not supported (Main Mic -> MICBIAS1)
[    2.943387] midas-audio sound: ASoC: no dapm match for Main Mic --> (null) --> MICBIAS1
[    2.960182] midas-audio sound: ASoC: Failed to add route Main Mic -> direct -> MICBIAS1
[    2.976927] midas-audio sound: Connecting non-supply widget to supply widget is not supported (Headset Mic -> MICBIAS2)
[    3.001818] midas-audio sound: ASoC: no dapm match for Headset Mic --> (null) --> MICBIAS2
[    3.026752] midas-audio sound: ASoC: Failed to add route Headset Mic -> direct -> MICBIAS2
[    3.051950] midas: midas_late_probe: snd_soc_dapm_force_enable_pin AIF1CLK ret = 0
[    3.170498] input: Midas Headset as /devices/platform/sound/sound/card0/input5
[    7.471188] midas_mic_id: detected jack: status=515
[   12.299236] midas_mic_id: detected jack: status=515
[   14.736722] midas_mic_id: detected jack: status=515
```

It fails to register card with errno 517, but as I've added a hack to force retry, it succeeds afterwards.
Log also shows that the headset device is created and when inserting the headphones into the jack, midas_mic_id callback returns 515 (in decimal, or in hex - 0x203). I'm not quite sure, what values 0x200 and 0x3 stands for. On stock this callback returns 0x403 (comment in the stock kernel sources says 0x400 means a high impedance and the microphone is detected: https://github.com/ReplicantOS-midas/android_kernel_samsung_smdk4412/commit/d202e1a49ae4df0f92a28c0a15d4ca3e1d67d424#diff-23f1489748c71781de9a885103c63514R338 ).

The problem is that routes Main Mic -> MICBIAS1 and  Headset Mic -> MICBIAS2 are not working, which I think could be a potential cause of why headphone is not detected.

**Comment by herbsmn** [comment]
Hello again @ChronoMonochrome,

You mentioned that if you get those two issues above resolved that the device is "almost usable". We are curious, if audio with headphones and deep sleep get fixed, do you plan on releasing any Roms for LineageOS? Or are there other issues that you'd like to see fixed before you would publish Rom images?

GNUtoo is focusing on libsamsung-ril, libsamsung-ipc, and the upstream kernel. It is our understanding, though, that you have a RIL adaptation layer to reuse the nonfree libraries for the RIL, which is something our project won't be using.

In Solidarity,
Kurtis

**Comment by GNUtoo** [comment]
Hi,

-517 is EPROBE_DEFER, it means that it's probably waiting for a missing driver or power supply regulator or something like that.

Are all the required regulators there in the devicetree?

**Comment by ChronoMonochrome** [comment]
> You mentioned that if you get those two issues above resolved that the device is "almost usable". We are curious, if audio with headphones and deep sleep get fixed, do you plan on releasing any Roms for LineageOS? Or are there other issues that you'd like to see fixed before you would publish Rom images?

Probably still not quite usable for the most of people at this moment (especially for those who are using S3 as a daily driver). Could be usable for that a minority who mostly use this phone as a player, but I'm probably not gonna release builds even if those issues are resolved (just too early to do that).

> GNUtoo is focusing on libsamsung-ril, libsamsung-ipc, and the upstream kernel. It is our understanding, though, that you have a RIL adaptation layer to reuse the nonfree libraries for the RIL, which is something our project won't be using.

Right now I don't have a working solution for a RIL even with a non-free blobs (actually I didn't try that out, don't know if it's easy to make work with the mainline kernel). But generally I'm just focusing on other issues than RIL.

> -517 is EPROBE_DEFER, it means that it's probably waiting for a missing driver or power supply regulator or something like that.

This error doesn't seem to be fatal. There are probably some issues with the order the stuff is loaded up in the DTS - because, if I wait a little bit, and then retry, the sound card is loaded up fine.
Was looking at the stock kernel sound driver, did a lot of simplifications (and even was able to get mainline version of the driver in 3.0 to work), but still didn't figure out the issue I was looking for. Headphones are working fine on 3.0 mainline version (with the legacy non-DTS setup used in 3.0).
If anyone wants to have a look at the 3.0 driver version, here is the link - https://github.com/ChronoMonochrome/android_kernel_samsung_smdk4412/commits/sound-wip-3.0
So for now I'm stuck again. I didn't look at the mfd (drivers/mfd/wm8994*) driver, which could be as well a culprit.
P.s. the main goal of messing up with 3.0 kernel is to hopefully trigger the same issue as in 5.3+ kernel, just to make it more clear.

**Comment by GNUtoo** [comment]
> Right now I don't have a working solution for a RIL even with a non-free blobs (actually I didn't try that out, don't know if it's easy to make work with the mainline kernel). But generally I'm just focusing on other issues than RIL.

I'm doing the opposite. Simon Shields did some early bringup work for the modem, and it was probably untested, so I used that as a base for my work.

Here's the current status:

I currently do minimal testing (modem bootstrap + wait for 2 ipc messages + reboot) under GNU/Linux as it's way faster to do it this way.

While the modem boots, there are still some minor things to implement (for instance I've to reboot the device between tests, but it's not a big issue for me as the boot time is almost negligible in my test setup).

I also need to fix various USB issues in the kernel:
- On 5.2 the full USB bus is reset when the HSIC is reset, but it does come back afterward.
- On 5.4 the USB bus doesn't come back afterward, which is really annoying, so I mostly use 5.2 for testing my work on libsamsung-ipc.

I've also still need to finish some big cleaning of the code.

With libsamsung-ipc, I've also some big cleanups to finish, and possibly implement some small missing things.

With libsamsung-ril, I need to convert it to the new SIM state API, and possibly to do some small cleanups.

Also, while the API is a bit different from the Samsung kernels (hence the work in libsamsung-ipc), it's also not in line with what is required to upstream a driver like that as it's still close to the Samsung one.

 For instance it uses a open/read/write API instead of a network API like with the N900 modem that is very similar (both have a specific network protocol that have part of its implementation tied to specific transport buses).

So once I'll have the code working well enough and readable and all that, the libsamsung-ipc and kernel patches might be pushed in some replicant-9 branch, as the kernel API will be subject to change.

As for samsung-ril, it's not really hardware specific or kernel API specific, as this is done in libsamsung-ipc, so I'll most probably push it all.

**Comment by herbsmn** [comment]
Replicant has been working on Replicant 10. https://git.replicant.us/contrib/replicant-9/device_samsung_i9305/log/?h=replicant-10

@ChronoMonochrome I noticed you've been doing some work lately with your repos. Care to update us on what you've been working on?

**Comment by ChronoMonochrome** [comment]
> Replicant has been working on Replicant 10. https://git.replicant.us/contrib/replicant-9/device_samsung_i9305/log/?h=replicant-10
> 

Nice to know!

> @ChronoMonochrome I noticed you've been doing some work lately with your repos. Care to update us on what you've been working on?

I'm still struggling with the stuff yet not supported on the mainline, like the extcon/jack support & USB host mode. Have been looking into the Tizen 3.10 kernel code, not completely sure if I can borrow some stuff from it, but this is what I'm trying to do. It's a pity I can't boot 3.10 kernel to test the stuff directly (there are some sort of display issues, so testing is somewhat difficult). I'm currently testing the stuff in Ubuntu 16.04.

**Comment by ChronoMonochrome** [comment]
@herbsmn 

I have been doing some experiments with alsactl in the buildroot with Linux 4.6 kernel.
It appears the jack works normally in buildroot, but for some reason doesn't work in Ubuntu and Android.

**Comment by herbsmn** [comment]
@javifo is doing some new work on upstream midas here: https://github.com/javifo/replicant_kernel_linux/commits/lk-5.3

**Comment by ChronoMonochrome** [comment]
@herbsmn 

that's nice! I see a commit that fixes the touchscreen issues, great, this bug was quite annoying.
Small update to the headset issue: I managed to get headset detection working with an ugly hack in the kernel: https://github.com/ReplicantOS-midas/android_kernel_samsung_smdk4412/commit/0c95d0d4104393d1b7812064bc47320030a069ea

What was necessary to do make it working is to toggle the mixer headphone switch manually.
I haven't found a way to detect the headset removal, so there are no other ways than polling the register in order to enable / disable the switch...

Ideally, this should not even be in the kernel, but somehow the userspace fails to do this work.

**Comment by herbsmn** [comment]
Our repo moved over to here: https://git.replicant.us/contrib/replicant-next

**Comment by herbsmn** [comment]
> I think the problem lies somewhere in kernel (DTS?), as it seems like the sound support for wm1811 is not fully implemented.

It looks like a samsung employee has been submitting midas related patches upstream, including a patch set for audio support: https://patchwork.kernel.org/project/linux-samsung-soc/list/?submitter=7616

It looks like you can test out the audio patch as of today using the sound-5.9-rc1 branch, as is explained here: https://www.mail-archive.com/linux-kernel@vger.kernel.org/msg2260833.html

Also, Replicant devs have been hacking a lot on graphics stuff, as you can see from the patches in replicant-next at the link in the post above.

Things are starting to come together quite nicely!

**Comment by ChronoMonochrome** [comment]
@herbsmn 
it's nice to see more stuff is getting accepted by the upstream!

I was recently trying to boot up Replicant 10 build on my i9300, but I can't seem to get graphics to work.

```
08-21 05:53:51.886  1827  1827 I ServiceManager: Waiting for service 'SurfaceFlinger' on '/dev/binder'...
08-21 05:53:51.913  1869  1869 E hwc-drm-device: Failed to set universal plane cap -1
08-21 05:53:51.913  1869  1869 E hwc-resource-manager: Failed to initialize any displays
08-21 05:53:51.913  1869  1869 E hwc-drm-two: Can't initialize the resource manager -22
08-21 05:53:51.913  1869  1869 E hwc-drm-two: Failed to initialize DrmHwcTwo err=6
08-21 05:53:51.913  1869  1869 E ComposerHal: failed to open hwcomposer device: Invalid argument
```
(with Linux 5.8-rc6 from https://git.replicant.us/contrib/replicant-next)

Any known resolution to this?
I've briefly checked graphics related changes and found this commit: https://git.replicant.us/contrib/replicant-next/device_samsung_midas_common/commit/?h=replicant-10&id=62d61abe032f700cd73a6eefddd498d8e8bc29be ("Override default gralloc service to prevent race condition with hwcomposer").
Despite of the presence of this commit it doesn't seem like hwcomposer works. 
I must've missing something, probably (the commit mentions the gralloc service needs to be started in init.rc, but I couldn't find if it's started anywhere).

Meanwhile, 5.9-rc1 kernel has some issues with the HWC too:

```
08-21 06:42:31.766   204   337 E hwc-platform-drm-generic: failed to import prime fd 12 ret=-1
08-21 06:42:31.766   204   337 E hwc-drm-two: Failed to import layer, ret=-1
08-21 06:42:31.766   216   216 E HWComposer: presentAndGetReleaseFences: present failed for display 0: NoResources (6)
08-21 06:42:31.778   495   495 D KeyguardClockSwitch: Updating clock: 6:42
08-21 06:42:31.782   204   337 E hwc-platform-drm-generic: failed to import prime fd 13 ret=-1
08-21 06:42:31.782   204   337 E hwc-drm-two: Failed to import layer, ret=-1
08-21 06:42:31.782   216   216 W HwcComposer: command 0x4010000 generated error 8
08-21 06:42:31.782   216   216 E RenderEngine: Failed to bind framebuffer! Aborting GPU composition for buffer (0xa72a9cb0).
08-21 06:42:31.782   204   337 E hwc-platform-drm-generic: failed to import prime fd 14 ret=-1
08-21 06:42:31.782   204   337 E hwc-drm-two: Failed to import layer, ret=-1
```

I will stick with Replicant 9 for the moment as I've tried different setups of the render nodes in build.prop but none seemed to work.

**Comment by ChronoMonochrome** [comment]
@javifo 

I've just noticed some wifi bringup related commits in your repositories.
I managed to get Wifi working on Replicant 9 recently (I took your initial changes and added some stuff from our LOS 16.0 device tree and it appeared to be enough to get this to work).

https://github.com/ReplicantOS-midas/android_device_midas_common/commits/replicant-9
https://github.com/ReplicantOS-midas/android_device_samsung_i9300/commits/replicant-9

Trying to get some more HALs to work...

**Comment by herbsmn** [comment]
Do you happen to have any updates on this work ChronoMonochrome?

**Comment by ChronoMonochrome** [comment]
@herbsmn 

I didn't work on these sources with 3.0.101 kernel for a while now.

I made some builds based on Replicant 11 sources here:

https://github.com/Midas-Mainline/android_device_midas_common

Supports Audio (still using hacks for the headset) Wifi, USB-OTG (with dirty hacks in kernel) and sensors.
As it was also mentioned on replicant, the graphics speed has improved by quite a lot and comparable to what we had with smdk4412 kernel based builds.
Don't know if I return to 3.0.101 kernel (I haven't done any updates to mainline-based builds for a while too, though).
I was planning to check if Camera HAL works with mainline kernel or investigate Modem initialization on smdk4412 kernel, but still haven't done anything of that.
Also, since the kernel now has USB-OTG support, I was wondering if some external modems could work? (have no idea if Android supports this, but probably should work with GNU Linux).

**Comment by herbsmn** [comment]
This thread, which includes @xc-racer99, has some tips and hints on how to potentially move forward with modem initialization. https://lists.osuosl.org/pipermail/replicant/2021-August/003531.html

**Comment by tlaurion** [comment]
/e/ seems to remove builds for nougat Midas based platforms... 0.18 is not supporting t0lte.

Seems like I will have to rebuild my kgpe-d16 build server...


Status updates?
Update: https://redmine.replicant.us/projects/replicant/wiki/PortingToAndroid11

Reading status there...

Edit: 
Status report says

- no u-boot
- no radio
- slow graphics

Or the page is not up to date?
Edit: https://redmine.replicant.us/projects/replicant/wiki/PortingToAndroid11/history
It was updated recently. 

https://redmine.replicant.us/projects/replicant/wiki/RunningReplicant11

Q: How is NlNet funded work progressing?

**Comment by tlaurion** [comment]
@GNUtoo @herbsmn @ChronoMonochrome @dllud @fourkbomb ?

**Comment by kekzoz** [comment]
@ChronoMonochrome I've been using your audio.i9300.xml configuration in a postmarketOS setup for t0lte.  The speaker works great but I get distortion on the headphones.  Have you (or anyone else) encountered this?  Any suggestions to fix it?  Thanks a lot.  Jack K

**Comment by ChronoMonochrome** [comment]
@kekzoz audio.i9300.xml was copied from LineageOS (with some small changes to make it work with a new HAL).
I would probably assume t0lte (and other devices) should use the corresponding audio config files (i.e. https://github.com/LineageOS/android_device_samsung_t0lte/blob/cm-14.1/configs/tiny_hw.xml).

**Comment by kekzoz** [comment]
@ChronoMonochrome  yes I tried using that file for my setup but headphones didn't work in that one either. I figured because you had headphone working using your i9300 file you might be able to shed some light on my issue. So you didn't encounter distorted audio in headphones?

**Comment by ChronoMonochrome** [comment]
@kekzoz  No, there were no such issues as the distortion, however, headphones didn't initially work at all on i9300. I tried fixing that in audio.i9300.xml file, but it seems like that didn't really work (can't recall now if my changes to that file were useful), as I ended up using an [ugly hack](https://github.com/Midas-Mainline/android_kernel_samsung_smdk4412/commit/784f676c90efb2c22978ba7e2cc572d2c23aaeea) in the kernel to manually switch to headphones when jack detects an accessory.
Maybe t0lte issue has something to do with the soundcard driver in the kernel, I recall that t0lte and i9300 had some differences in the sound driver in the vendor kernel.

**Comment by kekzoz** [comment]
@ChronoMonochrome - Ah that is helpful, thanks! I thought the hack referred to was just this one - https://github.com/Midas-Mainline/android_device_midas_common/commit/59f9b4f0cd2f1a409218551fb30212b9f5474a9b

**Comment by herbsmn** [comment]
Hi @ChronoMonochrome, did you succeed in getting rid of your ugly hack here? https://github.com/Midas-Mainline/android_kernel_samsung_smdk4412/commit/b0aa06f32213f3eab1ec9c429f64f126f9c8f2b2

I am also curious whether you'd be willing to comment on how well things seemed to work with LOS on the linux 5.18 kernel, which it looks like you were working on recently here: https://github.com/Midas-Mainline/android_kernel_samsung_smdk4412/tree/linux-5.18

**Comment by ChronoMonochrome** [comment]
Hi @herbsmn

> Hi @ChronoMonochrome, did you succeed in getting rid of your ugly hack here? https://github.com/Midas-Mainline/android_kernel_samsung_smdk4412/commit/b0aa06f32213f3eab1ec9c429f64f126f9c8f2b2


No, the hack is still there. I don't understand how the jack detection is supposed to work in userspace.
Unfortunately, I'm not super familiar with how the stuff is implemented ALSA on Ubuntu, and I couldn't find any information on the jack detection.
The only documentation I could find was from  [kernel.org](https://www.kernel.org/doc/html/v5.18/sound/soc/jack.html), but the jackdet code already works fine in the kernel, and it's never mentioned there how the userspace is supposed to use that.
I might need to add a bunch of debugging to the Audio HAL on i9300 if I ever wanna understand how the downstream kernel reports the jack detection to the HAL.

> I am also curious whether you'd be willing to comment on how well things seemed to work with LOS on the linux 5.18 kernel, which it looks like you were working on recently here: https://github.com/Midas-Mainline/android_kernel_samsung_smdk4412/tree/linux-5.18

I've made a kernel package for my custom installation of PostmarketOS there in June, but I haven't used my i9300 that often since then. 
Currently I have a triple-boot (Lineage 14.1 as the main system, PostmarketOS installed on the partitioned SDCard and Ubuntu image file also on SDCard).
PostmarketOS worked quite nicely with 5.18 kernel, I was hoping to get some new stuff working but unfortunately I didn't make much of progress here.

**Comment by herbsmn** [comment]
Thanks for the report back!

I assume you already saw this, but it says here, "in userspace the WiredAccessoryManager monitors the /sys/class/switch/h2w for uevents when the headphone is (un)plugged." https://wiki.postmarketos.org/wiki/Audio#Headphone_jack

@Viciouss, any chance you discovered how to get headphones to work in android userspace on your devices using the mainline kernel? I see you have been working on some things in this arena. https://github.com/search?q=user%3AViciouss+headset&type=code

**Comment by ChronoMonochrome** [comment]
> I assume you already saw this, but it says here, "in userspace the WiredAccessoryManager monitors the /sys/class/switch/h2w for uevents when the headphone is (un)plugged." https://wiki.postmarketos.org/wiki/Audio#Headphone_jack

Yes, I saw this page - when I checked the WiredAccessoryManager code linked from that page, I noticed, this was only present in Android KitKat sources and unused in later Android versions, e.g. Android 7.1.2. 

Interestingly enough, the downstream kernel has that switch/h2w sysfs interface, but I never managed to find any references to that in the userspace, so I'm not sure if it's even used.

The other note about udev rules looks interesting, not sure it was here when I first read this article (or maybe I somehow missed it, not sure), as it looks as the only feasible way to get jack detection work in the userspace with at least not such as a huge hack as I currently use right now.

EDIT:
Actually this WiredAccessoryManager  code is present in the LOS 14.1, my bad.
For some reason I assumed it was outdated, as the switch interface was in the staging directory in kernel, and was not present in the mainline. I guess this explains how the jack detection works on the downstream kernel, I just searched in the wrong places.

**Comment by Viciouss** [comment]
@herbsmn For the p4note this is done with the sec_jack driver:

https://github.com/Viciouss/samsung_p4note_kernel_backup/blob/GT-N8000_KK_Opensource/drivers/misc/sec_jack.c

Initialization and callbacks can be found here:

https://github.com/Viciouss/samsung_p4note_kernel_backup/blob/GT-N8000_KK_Opensource/arch/arm/mach-exynos/p4note-jack.c

It looks like the kernel part is only changing a single GPIO. I had no luck in mainline with this yet, I have only tried this once though and gave up rather quick. As the sec_jack is a switch driver, I guess userspace is listening for the jack update and then enabling the corresponding profile for ALSA to make this happen.
